jinja2/filters.py

```diff
@@ -567,6 +567,8 @@ def do_indent(
             rv += u'\n' + u'\n'.join(
                 indention + line if line else line for line in lines
             )
+            # Unescape any strings that were escaped due to concatenation
+            rv = Markup(rv.unescape())
```

Comment:
Is this the correct place to unescape? Also, calling `Markup` right after `unescape` seems like a no-op, and might fail if `rv` isn't already markup. `s` should probably be unescaped as soon as possible and only if it's `Markup`, then escaped again at the end if it was previously escaped or if autoescape is enabled in the current context.

Reply:
@davidism Yeah, I'm fixing now. I wasn't aware that there were cases where a non-Markup type would be passed to the filter, as this is my first time diving into jinja2 code.

Reply:
I'm not certain on this, I'd have to play around with it, but it needs to be checked.

Reply:
The idea to call `Markup()` again after unescaping is because it will give us a `Markup` instance with the escaped chars unescaped. My fix will contain this solution instead:

```python
if lines:
    rv += u'\n' + u'\n'.join(
        indention + line if line else line for line in lines
    )
    try:
        # Unescape any strings that were escaped due to concatenation
        rv = Markup(rv.unescape())
    except AttributeError:
        # s was not a Markup instance
        pass
```

This way, if `s` is not a `Markup` instance, then the `unescape()` triggers an `AttributeError`, and `rv` will be left unmodified.

Reply:
Here's another example of what I mean by calling `Markup()`. This is because `.unescape()` returns a unicode type.

```python
>>> from markupsafe import Markup
>>> x = Markup('foo ') + "'bar baz'"
>>> x
Markup(u'foo &#39;bar baz&#39;')
>>> Markup(x.unescape())
Markup(u"foo 'bar baz'")
>>>
```